src/NTP/Client.hs

```diff
@@ -232,71 +226,85 @@ handleNtpPacket cli packet = do
+        Left  (_, _, err)    ->
+            log cli Warning $ sformat ("Error while receiving time: "%shown) err
+        Right (_, _, packet) ->
+            handleNtpPacket cli packet `catchAll` handleE
+  where
+    handleE = log cli Warning . sformat ("Error while handle packet: "%shown)
 
 startReceive :: NtpMonad m => NtpClient m -> m ()
 startReceive cli = do
-    sockets <- liftIO . atomically . readTVar $ (ncSocket cli)
+    sockets <- liftIO . atomically . readTVar $ ncSocket cli
     case sockets of
-        (Just sock1, Just sock2) -> do
-            void $ fork $ doReceive sock1 cli `catchAll` handleE
-            doReceive sock2 cli `catchAll` handleE
-        (Just sock1, _) -> doReceive sock1 cli `catchAll` handleE
-        (_, Just sock2) -> doReceive sock2 cli `catchAll` handleE
-        _ -> log cli Warning $ "Unexpected state to start"
+        BothSock sIPv4 sIPv6 -> do
+            void $ fork $ runDoReceive True sIPv4
+            runDoReceive False sIPv6
+        IPv4Sock sIPv4 -> runDoReceive True sIPv4
+        IPv6Sock sIPv6 -> runDoReceive False sIPv6
   where
-    -- got error while receiving data, recreate socket
-    handleE e = do
+    runDoReceive isIPv4 sock = doReceive sock cli `catchAll` (handleE isIPv4 sock)
+    -- got error while receiving data, retrying in 5 sec
+    handleE isIPv4 sock e = do
         closed <- liftIO . readTVarIO $ ncClosed cli
         unless closed $ do
-            log cli Debug $ sformat ("Socket closed, recreating (reason: "%shown%")") e
-            sock <- mkSockets $ ncSettings cli
-            closed' <- liftIO . atomically $ do
-                writeTVar (ncSocket cli) sock
-                readTVar  (ncClosed cli)
-            -- extra check in case socket was closed by stopping client
-            -- while we recreated socket
-            unless closed' $
-                startReceive cli
+            log cli Debug $ sformat ("doReceive failed on socket"%shown%
+                                     ", reason: "%shown%
+                                     ", recreate socket in 5 sec") sock e
+            liftIO $ threadDelay (5 :: Second)
+            serveraddrs <- liftIO udpLocalAddresses
+            newSockMB <- liftIO $
+                if isIPv4 then createAndBindSock selectIPv4 serveraddrs
+                else createAndBindSock selectIPv6 serveraddrs
+            case fst <$> newSockMB of
+                Nothing      -> log cli Warning "Recreating of socket failed"
+                Just newSock
+                    | isIPv4 -> do
+                          liftIO . atomically $
+                              modifyTVar' (ncSocket cli) (flip mergeSockets $ IPv4Sock newSock)
+                          runDoReceive isIPv4 newSock
+                    | otherwise -> do
+                          liftIO . atomically $
+                              modifyTVar' (ncSocket cli) (flip mergeSockets $ IPv6Sock newSock)
+                          runDoReceive isIPv4 newSock
```

Comment:
Ah sorry, I was confused by `isIPv4`.
These rows differ only in `Ipv6Sock` part :/